File: server-py/repositories/settings_repository.py
from __future__ import annotations

from typing import Any
from database import get_db


def get_all() -> dict[str, str]:
    """读取所有设置项为 key-value 字典"""
    db = get_db()
    rows = db.execute("SELECT key, value FROM settings").fetchall()
    return {r["key"]: r["value"] for r in rows}


def upsert_all(settings: dict[str, str]) -> None:
    """批量写入或更新设置项

    ON CONFLICT(key) DO UPDATE 实现 upsert 语义：
    键已存在则更新 value，不存在则插入新行。
    with db: 确保所有操作在同一事务中执行。
    """
    db = get_db()
    with db:
        for key, value in settings.items():
            db.execute(
                "INSERT INTO settings (key, value) VALUES (?, ?) "
                "ON CONFLICT(key) DO UPDATE SET value = excluded.value",
                (key, value),
            )

File: server/repositories/settingsRepository.ts
import { getDb } from '../db.js';
import { RawSettings } from '../types.js';

// 读取所有设置为键值对
export function getAll(): RawSettings {
  const db = getDb();
  const rows = db.prepare('SELECT key, value FROM settings').all() as { key: string; value: string }[];
  const result: RawSettings = {};
  for (const row of rows) {
    result[row.key] = row.value;
  }
  return result;
}

// 批量写入/更新设置，使用事务保证原子性
export function upsertAll(settings: Record<string, string>): void {
  const db = getDb();
  const upsert = db.prepare(
    'INSERT INTO settings (key, value) VALUES (?, ?) ON CONFLICT(key) DO UPDATE SET value = excluded.value'
  );
  const transaction = db.transaction(() => {
    for (const [key, value] of Object.entries(settings)) {
      upsert.run(key, value);
    }
  });
  transaction();
}
